Fix `CollectIndexRequests` when both a join and a full scan uses an index

ggevay opened the pull request:
This fixes a silly bug in `prune_and_annotate_dataflow_index_imports` where it was not picking up a full scan if the index was also used in some other way. I moved recognizing full scans into the main traversal inside `collect_index_reqs_inner`. The loop which used to add full scans before this PR now shouldn't have any work to do, so it soft_panics if it still finds something to do.

The above soft panic revealed that ids occurring only in `sink_exports` were also being added by this loop before. Now I made adding them explicit in the `sink_exports` loop, and added a separate index usage type for this. I also clarified the `index_exports` loop. And I moved both of these loop to after the main loop that calls `CollectIndexRequests`.

I also adjusted the `IndexUsageType` printing a bit based on what we discussed with @aalexandrov and @teskje in yesterday's optimizer sync.

I've also addressed @aalexandrov's retrospective comments on the [first PR](https://github.com/MaterializeInc/materialize/pull/21119).

I also added a lot more tests.

### Motivation

  * This PR fixes a previously unreported bug.

### Tips for reviewer

It's best to review commit by commit, and hide whitespace changes.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
  - [x] Nightly: https://buildkite.com/materialize/nightlies/builds/3347
  - [x] Full SLT: https://buildkite.com/materialize/sql-logic-tests/builds/5654
  - [x] I also manually tested SUBSCRIBEs in various situations (many of these were probably already being tested in slts): on a view, mat view, source, table, and indexed versions of all of these.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

teskje commented on doc/user/content/transform-data/optimization.md:
```diff
@@ -55,10 +55,11 @@ EXPLAIN SELECT * FROM foo WHERE x = 42 AND y = 'hello';
 - If the `WHERE` clause constrains more fields than your index key includes, then the index might still provide some speedup, but it won't necessarily be optimal: In this case, the index lookup is performed using only those constraints that are included in the index key, and the rest of the constraints will be used to subsequently filter the result of the index lookup.
 - If `OR` is used and not all arguments constrain the same fields, create an index for the intersection of the constrained fields. For example, if you have `WHERE (x = 51 AND y = 'bbb') OR (x = 76 AND z = 9)`, create an index just on `x`.
 - If `OR` is used and its arguments constrain completely disjoint sets of fields (e.g. `WHERE x = 5 OR y = 'aaa'`), try to rewrite your query using a `UNION` (or `UNION ALL`), where each argument of the `UNION` has one of the original `OR` arguments.
+- If the same input is being used in a join as well as being constrained by equalities to literals, then an index can be used to speed up either the join (see below) or the literal equalities (see above). Which of these will perform better depends on data characteristics.
```
I feel like this point would better fit into a separate section, since it refers both to the other things in this section (equalities in WHERE clauses) and things in the next section (JOINs). Perhaps also having an example would be good to help people understand how different indexes influence the query plans and what the trade-offs are.

ggevay replied:
You are right. I'll do these improvements in a separate PR. (I'll create a separate doc PR anyway to add docs for explaining the usage types.)

aalexandrov reviewed (commented):
We need to decide whether it's OK to release this with duplicates in the "usage" lists.

I am in favor of removing duplicates and sorting the string entries in the list (especially as it seems to be only a minor code change).

ggevay commented:
Thanks for the reviews! Ok, I'll remove the duplicate usage type list entries.

ggevay commented on src/transform/src/dataflow.rs:
```diff
@@ -523,11 +553,19 @@ fn prune_and_annotate_dataflow_index_imports(
+            .map(|(exported_index_id, _)| exported_index_id)
+            .any(|exported_index_id| exported_index_id == index_id));
+
+        *usage_types = Some(Vec::new());
```
changed

ggevay commented:
I addressed the comments. @aalexandrov, could you please check the lists? I simply did
```separated(", ", usage_types.iter().sorted().dedup())```
in `fmt_text`.

aalexandrov reviewed (commented):
Looks good!
